nupic/research/frameworks/dendrites/utils.py

```diff
@@ -19,78 +19,251 @@
+    # pairs of columns; the resulting tensor will have shape (num_units,
+    # num_categories, num_categories)
+    l2_norm = percentage_activations.norm(p=2, dim=1)
+    l2_norm = l2_norm.unsqueeze(1).repeat((1, num_dendrites, 1))
+    percentage_activations /= l2_norm
+
+    overlap_matrix = torch.bmm(percentage_activations.transpose(1, 2),
+                               percentage_activations)
+    return overlap_matrix
+
+
+def dendrite_overlap(dendrite_segments, context_vectors, selection_criterion):
+    """
+    Returns a 1D torch tensor with shape (num_units,) where entry k gives the overlap
+    score for the dendrites of unit k in relation to the specified context veectors and
+    specified classes/categories. The overlap score for a single unit is computed via
+    the following procedure:
+        - for each class/category, compute the categorical distribution that gives the
+          fraction of instances for which each dendrite becomes active (this function
+          calls `percent_active_dendrites`),
+        - normalize each probability vector to have L2 norm equal to 1,
+        - compute all pairwise similarity scores (i.e., dot products) of dendrite
+          activations between all classes/categories,
+        - average the pairwise dot products.
+
+    :param dendrite_segments: `DendriteSegments` object
+    :param context_vectors: iterable of 2D torch tensors with shape (num_examples,
+                            dim_context) where each 2D tensor gives a batch of context
+                            vectors from the same category
+    :param selection_criterion: the criterion for selecting which dendrites become
+                                active; either "regular" (for `GatingDendriticLayer`)
+                                or "absolute" (for `AbsoluteMaxGatingDendriticLayer`)
+    """
+    overlap_matrix = dendrite_overlap_matrix(dendrite_segments, context_vectors,
+                                             selection_criterion)
+
+    num_units, num_categories, _ = overlap_matrix.size()
+
+    # The overlap score is simply the average of the off-diagonal entries of the
+    # overlap matrix; in the ideal case with no dendrite overlap, the overlap matrix is
+    # the identity matrix
+
+    # Since the overlap matrix is symmetric, we only consider the lower half, excluding
+    # the diagonal entries since they are all guaranteed to be 1
+
+    # Mask for a lower triangular matrix
+    ltril_mask = torch.ones((num_categories, num_categories))
+    ltril_mask = ltril_mask.tril(diagonal=-1)
+    ltril_mask = ltril_mask.unsqueeze(0).repeat((num_units, 1, 1))
+
+    overlap_score = (overlap_matrix * ltril_mask).sum(dim=(1, 2))
+    overlap_score /= (0.5 * num_categories * (num_categories - 1))
+    return overlap_score
+
+
+def dendrite_duty_cycle(dendrite_segments, context_vectors, selection_criterion):
+    """
+    Returns a 2D torch tensor with shape (number of units, number of dendrites) where
```

Comment:
We should be consistent about how we count the dendrites. I've been counting the segments, so the shape here would be `(number of units, number of segments)`. Any thoughts on which might be better?

Reply:
We may be referring to the same thing. By "dendrites", I'm referring here to a weight vector (and in this case, each unit has 10 dendrites). The naming is confusing because the `DendriteSegments` class makes it sound like a segment is a set of weights for the entire layer of units. If you propose that we call each weight vector a "segment", then I'm on board with that. In that case, a "dendrite" would just be a single weight scalar in that weight vector.

Reply:
I believe we were referring to the same thing. My proposal was to call each weight vector a "segment". I think it's less ambiguous this way. It's also more consistent as the `DendritesSegments` module has an attribute called `num_segements`.

Reply:
I have changed the terminology in the files here to be consistent with you. In the future, I think we may need to propagate these changes to other dendrite files too.

Reply:
Thanks. I'll keep an eye out for where this occurs.